### ISSUE TITLE
Need For speed III

### ISSUE THREAD

**Issue by spirillen** [funilrys/PyFunceble#221]
## Bug description
Running a full test of https://github.com/Clefspeare13/pornhosts/tree/master/submit_here takes no less than:

`Execution Time: 01:00:01:21.243098`

## Modifications under `.PyFunceble.yaml`

Non as I can recall it, beside `hosts: False`

## Reproduction

Run https://github.com/Clefspeare13/pornhosts/blob/master/toolbox/domain-test.sh inside the repo

## Expected behavior
I would personally expect this to have finished in less than a hour...

## Screenshots
![image](https://user-images.githubusercontent.com/44526987/108066545-c4269900-705f-11eb-8845-164f1adcb789.png)

![image](https://user-images.githubusercontent.com/44526987/108066597-d43e7880-705f-11eb-9240-5e4db9b1e4c6.png)

## Versions

**OS:** Ubuntu 20,04.1 + Conda
**Python Version:** 3.9.1
**PyFunceble Version:** 4.0.0b22.dev as 4.0.0b23.dev isn't released ti Pipy https://github.com/funilrys/PyFunceble/actions/runs/565050940

## Additional context
By washing the info log from other repos, it's clear it is the backend that is the slow one here, DNS lookup's are fast enough, but I also see a lot of failed WHOIS (keybase://chat/spirillen,funilrys/3190) and last tested 20 years ago...

**Comment by funilrys** [comment]
Is the auto continue active here?

**Comment by spirillen** [comment]
```
grep 'auto' ~/miniconda/envs/pyfunceble4/.PyFunceble.yaml
    autocontinue: false
```

**Comment by funilrys** [comment]
Just a bit of context, as I found out where the "latency" actually - if not mostly - happens. 

PyFunceble, because of:

- reported issues 
- discovery by contributors
- discovery of me and/or @mitchellkrogza  while doing some manual backtesting of the output of Py - Funceble (early around 2017-19)

is not only doing an `NS` lookup. Early on we were only doing some `NS` lookup. But today, it is not the case. 

In fact, if you look at the source code: 

https://github.com/funilrys/PyFunceble/blob/bd6262a25af55f052bd721da83bf4c94f45c8965/PyFunceble/checker/availability/base.py#L640 

you can see that PyFunceble (for sub-domain testing) first try to query:

1. The `NS` server-s
2. The `CNAME` record, if the `NS` record was not found.
3. The `DNAME` record, if the `NS` and `CNAME` records were not found.
4. The `A` record, if the `NS`, `CNAME`, and `DNAME` records were not found.
5. The `AAAA` record, if the `NS`, `CNAME`, `DNAME` and `A` records were not found.

As you can imagine, we are doing this systematically for each subject. But, that's not the real explanation of the issue.

The latency is happening, because when we lookup, we systematically look/check (at this time) through all provided DNS servers. 

So, if we take the order I gave with `NS`, `CNAME`, `DNAME`, `A` and `AAAA` , that gives us, a real slow down because we are looking/checking each record overall provided DNS server. 
I mostly did it for safety, as PyFunceble has to somehow curate the data it reads. But I'm willing to change this.

What about the solution? It's not "easy" because I see this as a risk but not impossible if I really want to. 

The first solution is to change the order of the records to query to:

1. `NS`
2. `CNAME`
3. `A`
3. `AAAA`
5. `DNAME`

The second solution is the loop over the given DNS Server only if one is not responding or is giving us an error back. Otherwise, we follow (after mixing their order) the first one.

I already implemented both solutions locally and a significant improvement can be noticed when using more than 4 test workers. 

Before I push the patch, is this clear enough @spirillen? Do you have something to add?

**Comment by spirillen** [comment]
I most say, I'm a bit surprised, because as I did understand it, then we cycled (load balanced) thought the DNS severs DNS1 to record 1, DNS2 to record 2 etc, why I have added ~20 dns recursors to the lists

The idea about working through the lists of DNS servers is a good safety, but I think a third possibility could be just as safe, 

First DNS, no positive reply for any record => ask Second on the list and so on, until one is giving a positive response. If none of the DNS servers gives an active response..... well test fails, nothing new there.

In short: At first positive response = records DNS = Active, let's move on with the next record.

A note about the query order, it is good for any third(`www.`) level records, but on second level (example.org) we should use NS, A, AAAA, CNAME/DNAME as the domain (example.) reraly are CNAME/DNAME's :)

**Comment by funilrys** [comment]
@spirillen, what you are describing as the third possibility looks like what PyFunceble is doing now. 

Let's say we give 10 DNS servers and want to test `example.org`.

What PyFunceble is doing is:

- Ask server 1 about NS, CNAME, DNAME, A, AAAA. If they all (records) fail (without any response), continue to server 2.
- Ask server 2 about NS, CNAME, DNAME, A, AAAA. If they all (records) fail (without any response), continue to server 3.
- Ask server 3 about NS, CNAME, DNAME, A, AAAA. If they all (records) fail (without any response), continue to server N.
- Ask server N about NS, CNAME, DNAME, A, AAAA. If they all (records) fail and there is no more server available flag `example.org` as INACTIVE.

In this case, as soon as one is giving a response for one of the records, just consider `example.org` as ACTIVE.


What I'm willing to do:
- Ask server 1 about NS, CNAME, DNAME, A, AAAA. 
  - If we get an empty for all (records), flag `example.org` as INACTIVE. 
  - If we get an error or server 1 is not reachable, ask server N.
- Ask server N about NS, CNAME, DNAME, A, AAAA.
  - If we get an empty for all (records), flag `example.org` as INACTIVE. 
  - If we get an error or server 1 is not reachable, ask server N.

Of course, PyFunceble has already a way to shuffle the server order but it is disabled and not available to the public for now. I may expose the switch to the public soon 🤔. It is only available to those who directly use the Python API. 
It might be a good idea to shuffle the order too with what I'm willing to do.

**Comment by spirillen** [comment]
I really can't decide, but I think we should leave it as is... simply to ensure the highest hit rate possible... that most be the most important

**Comment by funilrys** [comment]
For those who are following this thread, my previous commit introduces the following argument:

- `--follow-server-order`
- `--trust-dns-server`

## `--follow-server-order`

This argument can be used to force PyFunceble to randomize the order of the DNS server. Meaning that the order is randomized before any DNS lookup.

## `--trust-dns-server`

When active, it does what I previously described (here) as "What I'm willing to do".

**Comment by spirillen** [comment]
2 those who can't read the kaybase chatbetween @funilrys and I, I'll just spread the word of https://github.com/funilrys/PyFunceble/commit/248486e16b607a641dbd1dab9f7b836cee31386b 

The rewrite of how, the output to file is process now, have added a speed gain up to 35-40% on tests of https://github.com/funilrys/PyFunceble/issues/225

can't wait to test on https://github.com/mypdns/porn-records and https://github.com/Clefspeare13/pornhosts

**Comment by spirillen** [comment]
2 DNS Servers
First run
![image](https://user-images.githubusercontent.com/44526987/109512664-9d775200-7aa4-11eb-8c03-2d015ae00d61.png)

2. run
![image](https://user-images.githubusercontent.com/44526987/109514841-bc76e380-7aa6-11eb-8040-7a508a5ecda7.png)


All DNS Servers from https://www.mypdns.org/P41
First run
![image](https://user-images.githubusercontent.com/44526987/109512741-b2ec7c00-7aa4-11eb-9222-3fe3a5d813e2.png)

2. run
![image](https://user-images.githubusercontent.com/44526987/109519771-d7982200-7aab-11eb-96a3-2c5fc6db0616.png)


The unexpected here is, more DNS server (Load balancing) actually are adding significant more time to the test cycle, I would had expected it to be the other way around!!

**Comment by BARMOLEY007** [comment]
Пн, 22 февр. 2021, 0:49 spirillen <notifications@github.com>:

**Comment by funilrys** [comment]
@spirillen, keep in mind: If there is a timeout error or a socket error, we (internally) loop over the whole list.

**Comment by spirillen** [comment]
sure, but it shouldn't be the case, they should all be stable and reliable DNS servers.  But let's discuss the list in https://www.mypdns.org/P41
